src/index/sparse/sparse_inverted_index.h

```diff
@@ -186,6 +186,29 @@ class InvertedIndex {
         }
     }
 
+    std::vector<float>
+    GetAllDistances(const SparseRow<T>& query, float drop_ratio_search, const BitsetView& bitset) const {
+        if (query.size() == 0) {
+            return {};
+        }
+        std::vector<T> values(query.size());
+        for (size_t i = 0; i < query.size(); ++i) {
+            values[i] = std::abs(query[i].val);
+        }
+        auto pos = values.begin() + static_cast<size_t>(drop_ratio_search * values.size());
+        std::nth_element(values.begin(), pos, values.end());
+        auto q_threshold = *pos;
+        std::shared_lock<std::shared_mutex> lock(mu_);
+        auto distances = compute_all_distances(query, q_threshold);
+        for (size_t i = 0; i < distances.size(); ++i) {
+            if (bitset.empty() || !bitset.test(i)) {
+                continue;
+            }
+            distances[i] = 0.0f;
+        }
+        return distances;
```

Comment:
technically, the code above is not performance-optimal.
There is no need to compute distances inside `compute_all_distances` for rows with `!bitset.test(i)`.

Reply:
`compute_all_distances` proceeds column by column. We can't skip entire rows directly in `compute_all_distances`. We'd have more bitset tests if we want to skip the computes which is slower.

```cpp
std::vector<float>
    compute_all_distances2(const SparseRow<T>& q_vec, T q_threshold, const BitsetView& bitset) const {
        std::vector<float> scores(n_rows_internal(), 0.0f);
        for (size_t idx = 0; idx < q_vec.size(); ++idx) {
            auto [i, v] = q_vec[idx];
            if (v < q_threshold || i >= n_cols_internal()) {
                continue;
            }
            auto lut_it = inverted_lut_.find(i);
            if (lut_it == inverted_lut_.end()) {
                continue;
            }
            // TODO: improve with SIMD
            auto& lut = lut_it->second;
            for (size_t j = 0; j < lut.size(); j++) {
                auto [idx, val] = lut[j];
                // branching for each value, instead of each row
                if (bitset.empty() || !bitset.test(idx)) {
                    scores[idx] += v * float(val);
                }
            }
        }
        return scores;
    }
```